### PR TITLE
fix: filter out broken symlinks when linking installs

### LINKED ISSUE #550: Package installation fails if package contains invalid symlinks
### Code to reproduce the issue:

There's no repro now as I'm experimenting with pnpm to generate packages on the fly. One of such packages contains broken symlink.

### Expected behavior:

Installation should put symlink as-is, e.g. in a broken state. That's what npm does.

### Actual behavior:

Installation fails on `fs.stat` call:
```
 ! ENOENT: no such file or directory, stat '/Users/andreypopp/.pnpm-store/1/bin_prot#df0d82a5081cb7a7c82379d4762471fc6fc8b6e7/src/blit_stubs_xen.c'
```

### Additional information:

 - `node -v` prints: v6.9.3
 - OS X

### PR DESCRIPTION
**Pull request by andreypopp**
We use `fs.lstat()` so we can gather the info about the link itself then use
`fs.exists()` in case of symbolic link to test if symlink is valid. Skip if it's
not.

Fixes #550

### CODE REVIEW

File: src/fs/hardlinkDir.ts [hunk 1]
```diff
@@ -9,8 +9,15 @@ export default async function hardlinkDir(existingDir: string, newDir: string) {
     dirs
       .map(async (relativePath: string) => {
         const existingPath = path.join(existingDir, relativePath)
-        const newPath = path.join(newDir, relativePath);
-        const stat = await fs.stat(existingPath)
+        const newPath = path.join(newDir, relativePath)
```
Comment (andreypopp) [inline src/fs/hardlinkDir.ts:14]: I removed the `;` here, can bring it back but that seems more consistent w/o it.

File: src/fs/hardlinkDir.ts [hunk 2]
```diff
@@ -9,8 +9,15 @@ export default async function hardlinkDir(existingDir: string, newDir: string) {
...
+        const stat = await fs.lstat(existingPath)
```
Comment (andreypopp) [inline src/fs/hardlinkDir.ts:15]: `fs.lstat` instead of `fs.stat` here so we don't fail on broken symlinks.

File: src/fs/hardlinkDir.ts (hunk 1 above)
Comment (zkochan) [reply src/fs/hardlinkDir.ts:14]: yes, thanks

File: src/fs/hardlinkDir.ts [hunk 3]
```diff
@@ -9,8 +9,15 @@ export default async function hardlinkDir(existingDir: string, newDir: string) {
...
+        if (stat.isSymbolicLink()) {
+          // filter out broken symlinks
+          if (!await fs.exists(existingPath)) {
```
Comment (zkochan) [inline src/fs/hardlinkDir.ts]: maybe this logic could go to the safeLink function? It has safe in its name so I guess it would be logical

Comment (andreypopp) [reply src/fs/hardlinkDir.ts]: Ok, will do

Comment (zkochan) [reply src/fs/hardlinkDir.ts]: if this is an error, maybe we should log a warning? You can `import logger from 'pnpm-logger'` and use the `logger.warn` function

**Comment by andreypopp** [comment]
Alright, added a test as well.

**Comment by andreypopp** [comment]
Funny that docpress now fails because of that broken symlink. I'll remove it and leave it only in the tarbal.

File: src/fs/hardlinkDir.ts (hunk 3 above)
Comment (andreypopp) [reply src/fs/hardlinkDir.ts]: Added a warning.

File: src/fs/hardlinkDir.ts [hunk 4]
```diff
@@ -9,24 +11,37 @@ export default async function hardlinkDir(existingDir: string, newDir: string) {
     dirs
       .map(async (relativePath: string) => {
         const existingPath = path.join(existingDir, relativePath)
-        const newPath = path.join(newDir, relativePath);
-        const stat = await fs.stat(existingPath)
+        const newPath = path.join(newDir, relativePath)
+        const stat = await fs.lstat(existingPath)
+        if (stat.isSymbolicLink()) {
```
Comment (zkochan) [inline src/fs/hardlinkDir.ts]: This can be combined with the 3rd statement, right? `stat.isSymbolicLink() || stat.isFile() `

Comment (andreypopp) [reply src/fs/hardlinkDir.ts]: Yeah, sorry, I was going to bed back then already.

I usually prefer not to dedup unrelated branches of code. In this case this is of course arguable.

**Pull request merged**